### ISSUE TITLE
Inconsistent Dependency Injection for SwaggerUIOptions and SwaggerOptions Compared to SwaggerGenOptions

### ISSUE THREAD

**Issue by Vaccano** [domaindrivendev/Swashbuckle.AspNetCore#1829]
Dependency Injection for `SwaggerUIOptions` and `SwaggerOptions` is inconsistent with `SwaggerGenOptions`.

When setting up `SwaggerGenOptions` using `IConfigureNamedOptions`, it works as I would expect.  Meaning that the `Configure` method of the class implementing `IConfigureNamedOptions` is called, then the lambda from `AddSwaggerGen` is called with the `SwaggerGenOptions` object that was the modified in the `IConfigureNamedOptions.Configure` method..

But for `SwaggerUIOptions` and `SwaggerOptions` it is either one or the other.  If you provide a lambda, then the `IConfigureNamedOptions` is not called.

This makes it hard to create libraries that expand the settings.  It means that any such library has to take ownership of ALL the settings.  Or expose non-standard callbacks, etc.

Here is a walk through on how to repro this issue:

1. Create a basic API and add Swashbuckle.AspNetCore 5.5.1.
2. Create a class like this:

        public class ExtraSwaggerUiOptions : IConfigureNamedOptions<SwaggerUIOptions>
        {
            public void Configure(SwaggerUIOptions options)
            {
                    options.DocumentTitle = "Configure was Called";
            }

            public void Configure(string name, SwaggerUIOptions options)
            {
                     options.DocumentTitle = "Configure was Called";
            }
        }
3. Register it in `ConfigureServices` like this:

     services.AddTransient<IConfigureOptions<SwaggerUIOptions>, ExtraSwaggerUiOptions >();

4. In `Configure` update it to have the following code:

            app.UseSwaggerUI(options =>
            {
                options.SwaggerEndpoint("/swagger/v1/swagger.json", "OpenAPI Authentication Demo");
            });

Result: The DocumentTitle is not modified.
Expected: Both the DocumentTitle and the SwaggerEndpoint would be modified. (With the `Configure` method being called first.)

As I said above,  The "Expected" result is how `SwaggerGenOptions` works.  This issue is to try to get `SwaggerUIOptions` and `SwaggerOptions` to align with `SwaggerGenOptions`

**Comment by Vaccano** [comment]
I have worked around this issue by creating an extension method on `SwaggerUIOptions` that requires the developer to pre-inject the needed information.  Not very nice, but it could be worse.  

I added this comment:

     // NOTE: this does not use the options pattern (allowing injection of MyCompanyOpenApiOptions) because the Swashbuckle system has a bug with SwaggerUIOptions.
     // If the options system is used then all of the SwaggerUIOptions must be set in the class implementing the options pattern.  (Were as normally it will allow for the 
     // call to the delegate to augment what was set.)

I hope to be able to remove the comment someday, but the reality is that we rarely get to go back and fix things like this :(

**Comment by domaindrivendev** [comment]
Historically, Swashbuckle only provided one approach for configuring it's middleware components - specifically passing a "setup action" to the corresponding middleware config method (`UseSwagger`, `UseSwaggerUI` etc.). This allowed you to do everything in one place (i.e. register _and_ configure the middleware). However, it does, as you've pointed out, bypass the "Options" infrastructure. This is because middleware config methods extend `IApplicationBuilder` as opposed to `IServiceCollection` and as far as I'm aware there is no way to inject options through `IApplicationBuilder`.

However, to allow developers take advantage of the "Options" infrastructure, Swashbuckle now supports a second approach whereby you can register the middleware by calling the relevant method _without_ any parameters (e.g. `UseSwaggerUI()`) and then configure it by injecting the relevant options (e.g. `IOptions<SwaggerUiOptions>`) from the `ConfigureServices` method of your startup class.

So in summary - if you provide a setup action then you're opting for the old approach that bypasses the "Options" infrastructure, and if you omit it then you're opting for the new approach and can inject the options instead.

@Vaccano if you use the latter approach, then I think things should work for you as expected.

**Comment by Vaccano** [comment]
Unfortunately that does not work for what I need (but my workaround is fine).  Here is why it does not work (you can judge if it is a common enough use case to be worth time coding).

I am writing a NuGet package for my company that layers on top of yours.  It sets up Swashbuckle to work with my company's IDP and API Gateway.  

Basically I need to setup the OAuth 2 settings in the `UseSwaggerUI` method and override the `SwaggerIndex.html` via the `IndexStream` property.  (I also add a server in the `UseSwagger` method)  However, these setting will not be all the settings that a developer will need to apply to `UseSwaggerUI`.  

The way Swashbuckle is coded now, if I use the options pattern in my NuGet, then any settings that are applied in `UseSwaggerUI` by users of my NuGet will be ignored.  I would have to code an alternate, nonstandard way for them to set those settings.

With the current "one or the other" setup, I either have to take over all the settings for the `UseSwaggerUI` method (via the Options pattern) or not use the options pattern at all.

I have opted for the "not use the options pattern" route, as I don't want to limit my developers to only the settings that I would add to my NuGet via the options pattern (if I used it).

As I said above, `SwaggerGenOptions` works the way that the Options pattern is intended to work (allowing both to be used and merging the settings together).  But `SwaggerOptions` and `SwaggerUIOptions` are setup to be "one or the other".  That will not work for those making a NuGet.

But NuGets that depend on your NuGet and want to use the Options pattern may not be common enough to warrant the extra time to fix this.  And my workaround is not bad code, just not as shiny as it could be.

**Comment by domaindrivendev** [comment]
Re-opening as there's some underlying issues with the merged PR - in particular it breaks the following usecase to emit v2 and v3 versions of the Swagger/OpenAPI JSON side by side is now broken:

```csharp
app.UseSwagger(c =>
{
    c.RouteTemplate = "swagger/{documentName}/swagger.json";
    c.SerializeAsV2 = true;
});

app.UseSwagger(c =>
{
    c.RouteTemplate = "/swagger/{documentName}/openapi.json";
});
```

**Comment by domaindrivendev** [comment]
See b6c6919c8fc19394f767d6159bb6286036e2e843.

Now, the above usecase is supported as follows:

```csharp
app.UseSwagger(new SwaggerOptions
{
    RouteTemplate = "/swagger/{documentName}/swagger.json",
    SerializeAsV2 = true
});
app.UseSwagger(new SwaggerOptions
{
    RouteTemplate = "/swagger/{documentName}/openapi.json",
});
```
